### PR TITLE
feat(forms): Validator.pattern accepts a RegExp

### LINKED ISSUE #10150: Validator.pattern should accept a Regular Expression instead of a String
**I'm submitting a ...**  (check one with "x")
```
[* ] bug report
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Here's how I'm building a form input field
Regular expression pattern:
 regexPattern:string  is "/^\d{2}\/\d{2}\/\d{4}$/"

group[question.key] = ['', Validators.compose([Validators.required,Validators.pattern(regexPattern)])];
FormBuilder.group(group)

During Debugging I see:
regexPattern : "/^d{2}/d{2}/d{4}$/"
{ "pattern": { "requiredPattern": "^/^d{2}/d{2}/d{4}$/$", "actualValue": "01/01/2001" } }

**Expected/desired behavior**
As you can see the requiredPattern adds an extra ^ at the beginning and a $ at the end. Validators.pattern(String) should accept a Regexp not a String. The regexPattern set was "/^\d{2}\/\d{2}\/\d{4}$/" but as you can see it's value while debugging is obvsiously requiredPattern": "^/^d{2}/d{2}/d{4}$/$"

Validators.patterns(String) should instead accept a regular expression literal Validators.pattern(RegExp). It should have  the same input signature as str.match(regexp) in javascript https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match

**Reproduction of the problem**
If the current behavior is a bug or you can illustrate your feature request better with an example, please provide the steps to reproduce and if possible a minimal demo of the problem via https://plnkr.co or similar (you can use this template as a starting point: http://plnkr.co/edit/tpl:AvJOMERrnz94ekVua0u5).

**What is the expected behavior?**

**What is the motivation / use case for changing the behavior?**

**Please tell us about your environment:**

* **Angular version:** 2.0.0-rc.4
 
* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ] 
 
* **Language:** [TypeScript X.X]

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes https://github.com/angular/angular/issues/10150

### CODE REVIEW

File: modules/@angular/forms/src/validators.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@
  */
 
 import {OpaqueToken} from '@angular/core';
+import {isString} from '@angular/facade/src/lang';
```
Comment (pkozlowski-opensource) [inline modules/@angular/forms/src/validators.ts]: Please don't use facades, we are in the process of removing them

File: modules/@angular/forms/src/validators.ts [hunk 2]
```diff
@@ -95,16 +94,16 @@ export class Validators {
...
    * Validator that requires a control to match a regex to its value.
    */
-  static pattern(pattern: string): ValidatorFn {
+  static pattern(pattern: string|RegExp): ValidatorFn {
     return (control: AbstractControl): {[key: string]: any} => {
       if (isEmptyInputValue(control.value)) {
         return null;  // don't validate empty values to allow optional controls
       }
-      const regex = new RegExp(`^${pattern}$`);
+      const regex = typeof pattern === 'string' ? new RegExp(`^${pattern}$`) : <RegExp>pattern;
```
Comment (vicb) [inline modules/@angular/forms/src/validators.ts]: `<RegExp>` should not be needed

File: modules/@angular/forms/src/validators.ts [hunk 3]
```diff
@@ -95,16 +94,16 @@ export class Validators {
...
       const value: string = control.value;
-      return regex.test(value) ?
+      return regex && regex.test(value) ?
```
Comment (vicb) [inline modules/@angular/forms/src/validators.ts]: not sure of the `!regex` case ? 
Should it rather returns an empty `ValidatorFn` ? `(_) => null`

Comment (vicb) [inline modules/@angular/forms/src/validators.ts]: Add a test

File: modules/@angular/forms/src/validators.ts (hunk 2 above)
Comment (DzmitryShylovich) [inline modules/@angular/forms/src/validators.ts]: tsc complains

File: modules/@angular/forms/test/validators_spec.ts [hunk 4]
```diff
@@ -101,36 +101,54 @@ export function main() {
          () => { expect(Validators.pattern('null')(new FormControl(null))).toBeNull(); });
 
       it('should not error on valid strings',
-         () => { expect(Validators.pattern('[a-zA-Z ]*')(new FormControl('aaAA'))).toBeNull(); });
+         () => expect(Validators.pattern('[a-zA-Z ]*')(new FormControl('aaAA'))).toBeNull());
 
       it('should error on failure to match string', () => {
         expect(Validators.pattern('[a-zA-Z ]*')(new FormControl('aaa0'))).toEqual({
-          'pattern': {'requiredPattern': '^[a-zA-Z ]*$', 'actualValue': 'aaa0'}
+          'pattern': {'requiredPattern': '/^[a-zA-Z ]*$/', 'actualValue': 'aaa0'}
```
Comment (vicb) [inline modules/@angular/forms/test/validators_spec.ts]: Could you please remove this breaking change ? (we would have to delay the PR until 3.0). Thanks

File: modules/@angular/forms/src/validators.ts [hunk 5]
```diff
@@ -95,16 +94,25 @@ export class Validators {
...
     return (control: AbstractControl): {[key: string]: any} => {
       if (isEmptyInputValue(control.value)) {
         return null;  // don't validate empty values to allow optional controls
       }
-      const regex = new RegExp(`^${pattern}$`);
       const value: string = control.value;
       return regex.test(value) ?
           null :
-          {'pattern': {'requiredPattern': `^${pattern}$`, 'actualValue': value}};
+          {'pattern': {'requiredPattern': `${regexStr}`, 'actualValue': value}};
```
Comment (arkon) [inline modules/@angular/forms/src/validators.ts]: Why not just directly put `regexStr`?

Comment (DzmitryShylovich) [inline modules/@angular/forms/src/validators.ts]: It would introduce a breaking change https://github.com/angular/angular/pull/12323#discussion_r83561460

**Pull request merged**